File: app/src/main/java/wangdaye/com/geometricweather/settings/ConfigStore.java
package wangdaye.com.geometricweather.settings;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;

import androidx.annotation.Nullable;
import androidx.preference.PreferenceManager;

import java.util.Map;
import java.util.Set;

public class ConfigStore {

    private final SharedPreferences mPreferences;

    private ConfigStore(SharedPreferences sp) {
        mPreferences = sp;
    }

    public static ConfigStore getInstance(Context context) {
        return new ConfigStore(
                PreferenceManager.getDefaultSharedPreferences(context));
    }

    public static ConfigStore getInstance(Context context, String name) {
        return new ConfigStore(
                context.getSharedPreferences(name, Context.MODE_PRIVATE));
    }

    public Map<String, ?> getAll() {
        return mPreferences.getAll();
    }

    @Nullable
    public String getString(String key, @Nullable String defValue) {
        return mPreferences.getString(key, defValue);
    }

    @Nullable
    public Set<String> getStringSet(String key, @Nullable Set<String> defValues) {
        return mPreferences.getStringSet(key, defValues);
    }

    public int getInt(String key, int defValue) {
        return mPreferences.getInt(key, defValue);
    }

    public long getLong(String key, long defValue) {
        return mPreferences.getLong(key, defValue);
    }

    public float getFloat(String key, float defValue) {
        return mPreferences.getFloat(key, defValue);
    }

    public boolean getBoolean(String key, boolean defValue) {
        return mPreferences.getBoolean(key, defValue);
    }

    public boolean contains(String key) {
        return mPreferences.contains(key);
    }

    public Editor edit() {
        return new Editor(mPreferences);
    }

    public static class Editor {

        final SharedPreferences.Editor mEditor;

        @SuppressLint("CommitPrefEdits")
        Editor(SharedPreferences sp) {
            mEditor = sp.edit();
        }

        public Editor putString(String key, @Nullable String value) {
            mEditor.putString(key, value);
            return this;
        }

        public Editor putStringSet(String key, @Nullable Set<String> values) {
            mEditor.putStringSet(key, values);
            return this;
        }

        public Editor putInt(String key, int value) {
            mEditor.putInt(key, value);
            return this;
        }

        public Editor putLong(String key, long value) {
            mEditor.putLong(key, value);
            return this;
        }

        public Editor putFloat(String key, float value) {
            mEditor.putFloat(key, value);
            return this;
        }

        public Editor putBoolean(String key, boolean value) {
            mEditor.putBoolean(key, value);
            return this;
        }

        public Editor remove(String key) {
            mEditor.remove(key);
            return this;
        }

        public Editor clear() {
            mEditor.clear();
            return this;
        }

        public boolean commit() {
            return mEditor.commit();
        }

        public void apply() {
            mEditor.apply();
        }
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/settings/ConfigStore.kt
package wangdaye.com.geometricweather.settings

import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceManager

class ConfigStore private constructor(sp: SharedPreferences) {

    companion object {

        @JvmStatic
        fun getInstance(context: Context) = getInstance(context, null)

        @JvmStatic
        fun getInstance(context: Context, name: String? = null): ConfigStore {
            return ConfigStore(
                if (name == null) {
                    PreferenceManager.getDefaultSharedPreferences(context)
                } else {
                    context.getSharedPreferences(name, Context.MODE_PRIVATE)
                }
            )
        }
    }

    private val preferences = sp

    fun preload() {
        // do nothing.
    }

    fun getString(key: String, defValue: String?): String? {
        return preferences.getString(key, defValue)
    }

    fun getStringSet(key: String, defValues: Set<String?>?): Set<String>? {
        return preferences.getStringSet(key, defValues)
    }

    fun getInt(key: String, defValue: Int): Int {
        return preferences.getInt(key, defValue)
    }

    fun getLong(key: String, defValue: Long): Long {
        return preferences.getLong(key, defValue)
    }

    fun getFloat(key: String, defValue: Float): Float {
        return preferences.getFloat(key, defValue)
    }

    fun getBoolean(key: String, defValue: Boolean): Boolean {
        return preferences.getBoolean(key, defValue)
    }

    fun contains(key: String): Boolean {
        return preferences.contains(key)
    }

    fun edit(): Editor {
        return Editor(this)
    }

    class Editor internal constructor(host: ConfigStore) {

        private val editor = host.preferences.edit()

        fun putString(key: String, value: String?): Editor {
            editor.putString(key, value)
            return this
        }

        fun putStringSet(key: String, values: Set<String?>?): Editor {
            editor.putStringSet(key, values)
            return this
        }

        fun putInt(key: String, value: Int): Editor {
            editor.putInt(key, value)
            return this
        }

        fun putLong(key: String, value: Long): Editor {
            editor.putLong(key, value)
            return this
        }

        fun putFloat(key: String, value: Float): Editor {
            editor.putFloat(key, value)
            return this
        }

        fun putBoolean(key: String, value: Boolean): Editor {
            editor.putBoolean(key, value)
            return this
        }

        fun remove(key: String): Editor {
            editor.remove(key)
            return this
        }

        fun clear(): Editor {
            editor.clear()
            return this
        }

        fun apply() {
            editor.apply()
        }
    }
}